src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts

```diff
@@ -0,0 +1,231 @@
+      this.initEdit();
+    } else {
+      this.initAdd();
+    }
+  }
+
+  initAdd() {
+    ['password', 'confirmpassword'].forEach((controlName) =>
+      this.userForm.get(controlName).setValidators([Validators.required])
+    );
+  }
+
+  initEdit() {
+    ['password', 'confirmpassword'].forEach((controlName) =>
+      this.userForm.get(controlName).setValidators([])
+    );
+    this.disableForEdit();
+    this.route.params.subscribe((params: { username: string }) => {
+      const username = params.username;
+      this.userService.get(username).subscribe((userFormModel: UserFormModel) => {
+        this.response = _.cloneDeep(userFormModel);
+        this.setResponse(userFormModel);
+      });
+    });
+  }
+
+  disableForEdit() {
+    this.userForm.get('username').disable();
+  }
+
+  setResponse(response: UserFormModel) {
+    ['username', 'name', 'email', 'roles'].forEach((key) =>
+      this.userForm.get(key).setValue(response[key])
+    );
+  }
+
+  getRequest(): UserFormModel {
+    const userFormModel = new UserFormModel();
+    ['username', 'password', 'name', 'email', 'roles'].forEach(
+      (key) => (userFormModel[key] = this.userForm.get(key).value)
+    );
+    return userFormModel;
+  }
+
+  createAction() {
+    const userFormModel = this.getRequest();
+    this.userService.create(userFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `User "${userFormModel.username}" has been created.`,
+          'Create User'
+        );
+        this.router.navigate(['/users']);
+      },
+      () => {
+        this.userForm.setErrors({ cdSubmitButton: true });
+      }
+    );
+  }
+
+  editAction() {
+    if (this.isUserRemovingNeededRolePermissions()) {
+      const initialState = {
+        titleText: 'Update user',
+        buttonText: 'Continue',
+        bodyTpl: this.removeSelfUserReadUpdatePermissionTpl,
+        onSubmit: () => {
+          this.modalRef.hide();
+          this.doEditAction();
+        },
+        onCancel: () => {
+          this.userForm.setErrors({ cdSubmitButton: true });
+          this.userForm.get('roles').reset(this.userForm.get('roles').value);
+        }
+      };
+      this.modalRef = this.modalService.show(ConfirmationModalComponent, { initialState });
+    } else {
+      this.doEditAction();
+    }
+  }
+
+  private isCurrentUser(): boolean {
+    return this.authStorageService.getUsername() === this.userForm.get('username').value;
+  }
+
+  private isUserChangingRoles(): boolean {
+    const isCurrentUser = this.isCurrentUser();
+    return (
+      isCurrentUser &&
+      this.response &&
+      !_.isEqual(this.response.roles, this.userForm.getValue('roles'))
+    );
+  }
+
+  private isUserRemovingNeededRolePermissions(): boolean {
+    const isCurrentUser = this.isCurrentUser();
+    return isCurrentUser && !this.hasUserReadUpdatePermissions(this.userForm.get('roles').value);
```

Comment:
There are a view `this.userForm.get(xyz).value` in this component, which you can now replace with `this.userForm.getValue(xyz)`

Reply:
From `CdFormGroup.getValue` documentation:

>  Get the value of a control if it has none it will return false

Note that `hasUserReadUpdatePermissions` is expecting an array, that's why I would prefer to not user `CdFormGroup.getValue` for production code.

IMO we should change the implementation of `CdFormGroup.getValue` in the future, to not transform `none` value into `false` value.